src/core/settingmanager.cpp

```diff
@@ -143,6 +157,74 @@ void SettingManager::doNotTrack()
                                         m_doNotTrackConfItem->value(false));
 }
 
+void SettingManager::handleObserve(const QString &message, const QVariant &data)
+{
+    const QVariantMap dataMap = data.toMap();
+    if (message == QLatin1String("embed:search")) {
+        QString msg = dataMap.value("msg").toString();
+        if (msg == QLatin1String("init")) {
+            const StringMap configs(OpenSearchConfigs::getAvailableOpenSearchConfigs());
+            const QStringList configuredEngines = configs.keys();
+            QStringList registeredSearches(dataMap.value(QLatin1String("engines")).toStringList());
+            QString defaultSearchEngine = dataMap.value(QLatin1String("defaultEngine")).toString();
+
+            m_searchEnginesInitialized = !registeredSearches.isEmpty();
+
+            // Upon first start, engine doesn't know about the search engines.
+            // Engine load requests are send within the for loop below.
+            if (!m_searchEnginesInitialized) {
+                m_addedSearchEngines = new QStringList(configuredEngines);
```

Comment:
Yes there could but imo this is still better than having QStringList allocated to the stack.

As we're interested only about the first time when search engines are added (registeredSearches is empty) => ERROR_DUPLICATE_ENGINE (0x2) should not occur. 

As ```confirm``` argument is always false when adding search engines that rules out one ERROR_UNKNOWN_FAILURE. Didn't manage to create failure with local files. However, with something like ```http://invalid/data/engine.xml``` an error will be triggered.